File: LeetCode/207CourseSchedule/CourseSchedule.cpp
class Solution {
public:
    bool dfs(int course, map<int,vector<int>> &prereq, set<int> &visited) {
        if (visited.find(course) != visited.end()) {
            return false;
        }
        if (prereq[course].size() == 0) {
            return true;
        }
        visited.insert(course);
        for (int i = 0; i < prereq[course].size(); i++) {
            if (!dfs(prereq[course][i], prereq, visited)) {
                return false;
            }
        }
        visited.erase(course);
        prereq[course].clear();
        return true;
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // DFS -> Time:O(n+p), Space:O(p)
        map<int,vector<int>> prereq;
        set<int> visited;
        for (int i = 0; i < prerequisites.size(); i++) {
            if (prereq.find(prerequisites[i][0]) == prereq.end()) {
                vector<int> p;
                prereq[prerequisites[i][0]] = p;
            }
            prereq[prerequisites[i][0]].push_back(prerequisites[i][1]);
        }
        for (int i = 0; i < numCourses; i++) {
            if (!dfs(i, prereq, visited)) {
                return false;
            }
        }
        return true;
    }
};

File: LeetCode/207CourseSchedule/CourseSchedule.py
lass Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # DFS -> Time:O(n+p), Space:O(p)
        prereq = {i:[] for i in range(numCourses)}
        visited = set()

        for crs, pre in prerequisites:
            prereq[crs].append(pre)
        
        def dfs(course):
            if course in visited:
                return False
            if not prereq[course]:
                return True
            visited.add(course)
            for crs in prereq[course]:
                if not dfs(crs):
                    return False
            visited.remove(course)
            prereq[course] = []
            return True

        for i in range(numCourses):
            if not dfs(i):
                return False
        return True
